File: desktop/sshconfig.go
package main

import (
	"bufio"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// SSHConfigHost is one named host parsed from ~/.ssh/config — a candidate
// the New SSH Project dialog can pre-fill from. Name is the alias the user
// wrote after `Host`; HostName is the actual address from the optional
// `HostName` directive (falls back to Name when absent).
type SSHConfigHost struct {
	Name         string `json:"name"`
	HostName     string `json:"hostName"`
	User         string `json:"user"`
	Port         int    `json:"port"`
	IdentityFile string `json:"identityFile"`
}

// ListSSHHosts returns non-wildcard hosts from ~/.ssh/config (and any
// `Include`d files) for the dialog's host picker. A missing config file is
// not an error — the picker just shows nothing.
func (a *App) ListSSHHosts() ([]SSHConfigHost, error) {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		return []SSHConfigHost{}, nil
	}
	hosts, err := parseSSHConfig(filepath.Join(home, ".ssh", "config"), home, 0)
	if err != nil {
		return nil, err
	}
	hosts = dedupeSSHHosts(hosts)
	sort.Slice(hosts, func(i, j int) bool { return hosts[i].Name < hosts[j].Name })
	if hosts == nil {
		hosts = []SSHConfigHost{}
	}
	return hosts, nil
}

const sshIncludeMaxDepth = 4

func parseSSHConfig(path, home string, depth int) ([]SSHConfigHost, error) {
	if depth > sshIncludeMaxDepth {
		return nil, nil
	}
	f, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	defer f.Close()

	var (
		hosts   []SSHConfigHost
		blocks  []*sshConfigBlock
		current *sshConfigBlock
		// We can't evaluate `Match` predicates without faking ssh's
		// connection state, so any block under `Match` is ignored until the
		// next `Host`.
		skipMatch bool
	)

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		key, val, ok := parseSSHConfigLine(scanner.Text())
		if !ok {
			continue
		}
		switch strings.ToLower(key) {
		case "host":
			skipMatch = false
			names := []string{}
			for _, n := range splitSSHFields(val) {
				if !isSSHWildcardHost(n) {
					names = append(names, n)
				}
			}
			if len(names) == 0 {
				current = nil
				continue
			}
			current = &sshConfigBlock{names: names}
			blocks = append(blocks, current)
		case "match":
			skipMatch = true
			current = nil
		case "include":
			for _, pat := range splitSSHFields(val) {
				matches, _ := filepath.Glob(expandSSHIncludePath(pat, home))
				for _, m := range matches {
					sub, err := parseSSHConfig(m, home, depth+1)
					if err == nil {
						hosts = append(hosts, sub...)
					}
				}
			}
		default:
			if skipMatch || current == nil {
				continue
			}
			switch strings.ToLower(key) {
			case "user":
				if current.user == "" {
					current.user = strings.TrimSpace(val)
				}
			case "port":
				if current.port == 0 {
					if n, err := strconv.Atoi(firstSSHField(val)); err == nil && n > 0 {
						current.port = n
					}
				}
			case "identityfile":
				if current.identityFile == "" {
					current.identityFile = firstSSHField(val)
				}
			case "hostname":
				if current.hostName == "" {
					current.hostName = firstSSHField(val)
				}
			}
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}

	for _, b := range blocks {
		for _, name := range b.names {
			hosts = append(hosts, SSHConfigHost{
				Name:         name,
				HostName:     b.hostName,
				User:         b.user,
				Port:         b.port,
				IdentityFile: b.identityFile,
			})
		}
	}
	return hosts, nil
}

type sshConfigBlock struct {
	names        []string
	hostName     string
	user         string
	port         int
	identityFile string
}

// parseSSHConfigLine returns the keyword and value for a config line,
// stripping comments and the optional `=` separator. Returns ok=false for
// blanks and comment-only lines.
func parseSSHConfigLine(line string) (key, val string, ok bool) {
	line = strings.TrimSpace(line)
	if line == "" || strings.HasPrefix(line, "#") {
		return "", "", false
	}
	idx := strings.IndexAny(line, " \t=")
	if idx < 0 {
		return "", "", false
	}
	key = line[:idx]
	rest := strings.TrimSpace(line[idx:])
	rest = strings.TrimPrefix(rest, "=")
	return key, strings.TrimSpace(rest), true
}

// splitSSHFields tokenises a value, honouring double-quoted spans so paths
// or hostnames containing spaces survive intact.
func splitSSHFields(s string) []string {
	var out []string
	var cur strings.Builder
	inQuotes := false
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case c == '"':
			inQuotes = !inQuotes
		case (c == ' ' || c == '\t') && !inQuotes:
			if cur.Len() > 0 {
				out = append(out, cur.String())
				cur.Reset()
			}
		default:
			cur.WriteByte(c)
		}
	}
	if cur.Len() > 0 {
		out = append(out, cur.String())
	}
	return out
}

func firstSSHField(s string) string {
	parts := splitSSHFields(s)
	if len(parts) == 0 {
		return ""
	}
	return parts[0]
}

func isSSHWildcardHost(name string) bool {
	if name == "" {
		return true
	}
	if strings.HasPrefix(name, "!") {
		return true
	}
	return strings.ContainsAny(name, "*?")
}

func expandSSHIncludePath(pat, home string) string {
	if filepath.IsAbs(pat) {
		return pat
	}
	if strings.HasPrefix(pat, "~/") {
		return filepath.Join(home, pat[2:])
	}
	return filepath.Join(home, ".ssh", pat)
}

// dedupeSSHHosts keeps the first occurrence of each name (matches ssh's
// first-match-wins precedence for repeated Host blocks).
func dedupeSSHHosts(in []SSHConfigHost) []SSHConfigHost {
	seen := make(map[string]struct{}, len(in))
	out := make([]SSHConfigHost, 0, len(in))
	for _, h := range in {
		if _, dup := seen[h.Name]; dup {
			continue
		}
		seen[h.Name] = struct{}{}
		out = append(out, h)
	}
	return out
}

File: desktop/frontend/src-tauri/src/sshconfig.rs
// SSH host discovery — port of desktop/sshconfig.go. Reads ~/.ssh/config and
// any Include'd files, returning non-wildcard Host aliases for the Add-SSH-
// project picker. Missing config is not an error (returns []).
use serde::Serialize;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};

#[derive(Serialize, Clone)]
pub struct SshConfigHost {
    pub name: String,
    #[serde(rename = "hostName")]
    pub host_name: String,
    pub user: String,
    pub port: i64,
    #[serde(rename = "identityFile")]
    pub identity_file: String,
}

const SSH_INCLUDE_MAX_DEPTH: usize = 4;

#[tauri::command]
pub fn list_ssh_hosts() -> Result<Vec<SshConfigHost>, String> {
    let Some(home) = dirs::home_dir().filter(|h| !h.as_os_str().is_empty()) else {
        return Ok(vec![]);
    };
    let mut hosts = parse_ssh_config(&home.join(".ssh").join("config"), &home, 0)?;
    dedupe_hosts(&mut hosts);
    hosts.sort_by(|a, b| a.name.cmp(&b.name));
    Ok(hosts)
}

#[derive(Default)]
struct Block {
    names: Vec<String>,
    host_name: String,
    user: String,
    port: i64,
    identity_file: String,
}

fn parse_ssh_config(path: &Path, home: &Path, depth: usize) -> Result<Vec<SshConfigHost>, String> {
    if depth > SSH_INCLUDE_MAX_DEPTH {
        return Ok(vec![]);
    }
    let file = match std::fs::File::open(path) {
        Ok(f) => f,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(vec![]),
        Err(e) => return Err(e.to_string()),
    };

    let mut hosts: Vec<SshConfigHost> = vec![];
    let mut blocks: Vec<Block> = vec![];
    let mut current: Option<usize> = None;
    // `Match` predicates can't be evaluated without ssh's connection state, so
    // skip any block under `Match` until the next `Host`.
    let mut skip_match = false;

    for line in BufReader::new(file).lines() {
        let line = line.map_err(|e| e.to_string())?;
        let Some((key, val)) = parse_line(&line) else {
            continue;
        };
        match key.to_lowercase().as_str() {
            "host" => {
                skip_match = false;
                let names: Vec<String> = split_fields(&val)
                    .into_iter()
                    .filter(|n| !is_wildcard_host(n))
                    .collect();
                if names.is_empty() {
                    current = None;
                    continue;
                }
                blocks.push(Block {
                    names,
                    ..Block::default()
                });
                current = Some(blocks.len() - 1);
            }
            "match" => {
                skip_match = true;
                current = None;
            }
            "include" => {
                for pat in split_fields(&val) {
                    let glob_pat = expand_include_path(&pat, home);
                    let Ok(paths) = glob::glob(&glob_pat.to_string_lossy()) else {
                        continue;
                    };
                    for m in paths.flatten() {
                        if let Ok(sub) = parse_ssh_config(&m, home, depth + 1) {
                            hosts.extend(sub);
                        }
                    }
                }
            }
            other => {
                if skip_match {
                    continue;
                }
                let Some(idx) = current else { continue };
                let b = &mut blocks[idx];
                match other {
                    "user" if b.user.is_empty() => b.user = val.trim().to_string(),
                    "port" if b.port == 0 => {
                        if let Ok(n) = first_field(&val).parse::<i64>() {
                            if n > 0 {
                                b.port = n;
                            }
                        }
                    }
                    "identityfile" if b.identity_file.is_empty() => {
                        b.identity_file = first_field(&val);
                    }
                    "hostname" if b.host_name.is_empty() => {
                        b.host_name = first_field(&val);
                    }
                    _ => {}
                }
            }
        }
    }

    for b in &blocks {
        for name in &b.names {
            hosts.push(SshConfigHost {
                name: name.clone(),
                host_name: b.host_name.clone(),
                user: b.user.clone(),
                port: b.port,
                identity_file: b.identity_file.clone(),
            });
        }
    }
    Ok(hosts)
}

/// (key, value); None for blank/comment lines or lines without a separator.
fn parse_line(line: &str) -> Option<(String, String)> {
    let line = line.trim();
    if line.is_empty() || line.starts_with('#') {
        return None;
    }
    let idx = line.find([' ', '\t', '='])?;
    let key = line[..idx].to_string();
    let rest = line[idx..].trim_start();
    let rest = rest.strip_prefix('=').unwrap_or(rest);
    Some((key, rest.trim().to_string()))
}

/// Splits on whitespace, honouring double-quoted spans.
fn split_fields(s: &str) -> Vec<String> {
    let mut out = vec![];
    let mut cur = String::new();
    let mut in_quotes = false;
    for c in s.chars() {
        match c {
            '"' => in_quotes = !in_quotes,
            ' ' | '\t' if !in_quotes => {
                if !cur.is_empty() {
                    out.push(std::mem::take(&mut cur));
                }
            }
            _ => cur.push(c),
        }
    }
    if !cur.is_empty() {
        out.push(cur);
    }
    out
}

fn first_field(s: &str) -> String {
    split_fields(s).into_iter().next().unwrap_or_default()
}

fn is_wildcard_host(name: &str) -> bool {
    name.is_empty() || name.starts_with('!') || name.contains(['*', '?'])
}

fn expand_include_path(pat: &str, home: &Path) -> PathBuf {
    let p = Path::new(pat);
    if p.is_absolute() {
        return p.to_path_buf();
    }
    if let Some(rest) = pat.strip_prefix("~/") {
        return home.join(rest);
    }
    home.join(".ssh").join(pat)
}

/// Keep the first occurrence of each Host name (ssh's first-match-wins).
fn dedupe_hosts(hosts: &mut Vec<SshConfigHost>) {
    let mut seen = std::collections::HashSet::new();
    hosts.retain(|h| seen.insert(h.name.clone()));
}
